File: app/ios/Runner/VoiceDialogPlugin.swift
import Flutter
import Foundation
import AVFoundation

#if canImport(SpeechEngineToB)
import SpeechEngineToB

/**
 * Flutter ↔ 火山引擎 SpeechEngine Dialog SDK 桥接（iOS）
 *
 * 接入路径完全对齐官方文档（端到端 iOS SDK 接口文档）与 Android 端 VoiceDialogPlugin.kt：
 *   1. AppDelegate 启动时调用 `SpeechEngine.prepareEnvironment()`（仅一次）
 *   2. 创建引擎 → 设置参数（含 dialog address/uri）→ initEngine
 *   3. `SyncStopEngine` → `StartEngine(payload)` 一步到位
 *   4. 回调按 SEEngineStart / SEEngineStop / SEEngineError / 3000 段 Event 处理
 */
class VoiceDialogPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    static let methodChannelName = "com.starpath/voice_dialog"
    static let eventChannelName  = "com.starpath/voice_dialog_events"

    /// 官方固定值（文档 6561/1597643）
    private static let dialogAddress = "wss://openspeech.bytedance.com"
    private static let dialogUri     = "/api/v3/realtime/dialogue"
    /// X-Api-App-Key 为端到端实时语音服务固定值（非控制台 Secret Key）
    private static let fixedRealtimeDialogAppKey = "PlgvMymc7f3tQnJ6"

    private var eventSink: FlutterEventSink?
    private var engine: SpeechEngine?
    private var engineStarted = false

    /// 对齐 Android：StartEngine payload 里要带的模型 / 音色
    private var dialogModel: String = "1.2.1.1"
    private var ttsSpeaker: String  = "zh_female_vv_jupiter_bigtts"
    /// 防止 aiSpeaking 重复下发（先 CHAT_RESPONSE delta 到 → 再 TTS_SENTENCE_START）
    private var aiSpeakingEmitted = false

    static func register(with registrar: FlutterPluginRegistrar) {
        let instance = VoiceDialogPlugin()

        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - MethodChannel

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "prepareEnvironment":
            // AppDelegate 已在启动时调过；这里幂等返回，纯粹兼容老调用方。
            result(nil)
        case "startDialog":
            guard let args = call.arguments as? [String: Any] else {
                result(FlutterError(code: "MISSING_ARGS", message: "args required", details: nil))
                return
            }
            startDialog(args: args, result: result)
        case "stopDialog":
            stopDialog(result: result)
        case "interrupt":
            interrupt(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - startDialog

    private func startDialog(args: [String: Any], result: @escaping FlutterResult) {
        guard let appId    = args["appId"]    as? String,
              let appToken = args["appToken"] as? String else {
            result(FlutterError(code: "MISSING_PARAM", message: "appId/appToken required", details: nil))
            return
        }
        let resourceId = args["resourceId"] as? String ?? "volc.speech.dialog"
        self.dialogModel = (args["dialogModel"] as? String) ?? "1.2.1.1"
        self.ttsSpeaker  = (args["ttsSpeaker"]  as? String) ?? "zh_female_vv_jupiter_bigtts"
        let enableAec    = (args["enableAec"]   as? Bool)   ?? false

        // 录放并用 + 外放 + 蓝牙耳机兼容
        configureAudioSession()

        AVAudioSession.sharedInstance().requestRecordPermission { [weak self] granted in
            guard let self = self else { return }
            guard granted else {
                DispatchQueue.main.async {
                    result(FlutterError(code: "MIC_DENIED", message: "Microphone permission denied", details: nil))
                }
                return
            }
            DispatchQueue.main.async {
                self.initAndStart(appId: appId, appToken: appToken, resourceId: resourceId,
                                  enableAec: enableAec, result: result)
            }
        }
    }

    private func configureAudioSession() {
        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(
                .playAndRecord,
                mode: .voiceChat,
                options: [.defaultToSpeaker, .allowBluetooth, .allowBluetoothA2DP, .mixWithOthers]
            )
            try session.setActive(true, options: [])
        } catch {
            NSLog("[VoiceDialog] AVAudioSession configure failed: \(error)")
        }
    }

    /// 初始化引擎并用官方推荐的 `SyncStopEngine` + `StartEngine(payload)` 启动
    private func initAndStart(appId: String, appToken: String, resourceId: String,
                              enableAec: Bool, result: @escaping FlutterResult) {
        NSLog("[VoiceDialog] initEngine appId=\(appId.prefix(4))*** resourceId=\(resourceId) aec=\(enableAec) model=\(dialogModel) speaker=\(ttsSpeaker)")

        // 已有残留引擎先清掉（避免反复进入聊天页叠加）
        if let old = engine {
            _ = old.send(SEDirectiveSyncStopEngine)
            old.destroy()
            engine = nil
            engineStarted = false
        }

        engine = SpeechEngine()
        guard engine?.createEngine(with: self) == true else {
            engine = nil
            NSLog("[VoiceDialog] createEngineWithDelegate failed")
            result(FlutterError(code: "CREATE_ERROR", message: "createEngineWithDelegate failed", details: nil))
            return
        }

        // ── 必填参数（严格按官方文档）────────────────────────────────────────
        engine?.setStringParam(SE_DIALOG_ENGINE,                           forKey: SE_PARAMS_KEY_ENGINE_NAME_STRING)
        engine?.setStringParam(appId,                                      forKey: SE_PARAMS_KEY_APP_ID_STRING)
        engine?.setStringParam(VoiceDialogPlugin.fixedRealtimeDialogAppKey, forKey: SE_PARAMS_KEY_APP_KEY_STRING)
        engine?.setStringParam(appToken,                                   forKey: SE_PARAMS_KEY_APP_TOKEN_STRING)
        engine?.setStringParam(resourceId,                                 forKey: SE_PARAMS_KEY_RESOURCE_ID_STRING)
        engine?.setStringParam("starpath-user",                            forKey: SE_PARAMS_KEY_UID_STRING)
        // Dialog 网关地址 / URI（官方文档要求显式设置）
        engine?.setStringParam(VoiceDialogPlugin.dialogAddress,            forKey: SE_PARAMS_KEY_DIALOG_ADDRESS_STRING)
        engine?.setStringParam(VoiceDialogPlugin.dialogUri,                forKey: SE_PARAMS_KEY_DIALOG_URI_STRING)
        // ASR 流式（开启后才有 partial）
        engine?.setBoolParam(true, forKey: SE_PARAMS_KEY_ASR_SHOW_UTTER_BOOL)

        // ── AEC 可选 ────────────────────────────────────────────────────────
        if enableAec, let modelPath = Bundle.main.path(forResource: "aec_model", ofType: nil) {
            engine?.setBoolParam(true, forKey: SE_PARAMS_KEY_ENABLE_AEC_BOOL)
            engine?.setStringParam(modelPath, forKey: SE_PARAMS_KEY_AEC_MODEL_PATH_STRING)
        }

        // ── 初始化 ──────────────────────────────────────────────────────────
        let code = engine?.initEngine() ?? SENoError
        guard code == SENoError else {
            NSLog("[VoiceDialog] initEngine returned \(code.rawValue)")
            engine?.destroy()
            engine = nil
            result(FlutterError(code: "INIT_ERROR", message: "initEngine returned \(code.rawValue)", details: nil))
            return
        }
        NSLog("[VoiceDialog] initEngine OK")

        // ── 官方启动流程：SyncStop → StartEngine(payload) ───────────────────
        _ = engine?.send(SEDirectiveSyncStopEngine)

        let payload = buildStartEnginePayload()
        NSLog("[VoiceDialog] StartEngine payload=\(payload)")
        let startRet = engine?.send(SEDirectiveStartEngine, data: payload) ?? SENoError
        guard startRet == SENoError else {
            NSLog("[VoiceDialog] send(StartEngine) returned \(startRet.rawValue)")
            engine?.destroy()
            engine = nil
            result(FlutterError(code: "START_ERROR",
                                message: "send(StartEngine) returned \(startRet.rawValue)",
                                details: nil))
            return
        }

        // connected 事件在 SEEngineStart 回调里异步下发；这里先回调 Dart 侧 true。
        result(true)
    }

    /// StartEngine payload（与 Android buildStartEnginePayload 完全一致）
    /// - dialog.bot_name：人设名，默认豆包
    /// - dialog.extra.model：模型版本（O2.0 → 1.2.1.1，SC2.0 → 2.2.0.0）
    /// - asr.extra / tts.audio_config 不可为 null
    private func buildStartEnginePayload() -> String {
        let payload: [String: Any] = [
            "dialog": [
                "bot_name": "豆包",
                "extra": ["model": dialogModel]
            ],
            "asr": ["extra": [:]],
            "tts": [
                "audio_config": [:],
                "speaker": ttsSpeaker
            ]
        ]
        if let data = try? JSONSerialization.data(withJSONObject: payload, options: []),
           let s = String(data: data, encoding: .utf8) {
            return s
        }
        return "{}"
    }

    // MARK: - stopDialog / interrupt

    private func stopDialog(result: FlutterResult) {
        _ = engine?.send(SEDirectiveSyncStopEngine)
        engine?.destroy()
        engine = nil
        engineStarted = false
        aiSpeakingEmitted = false
        pushEvent(["type": "disconnected"])
        result(nil)
    }

    private func interrupt(result: FlutterResult) {
        _ = engine?.send(SEDirectiveEventClientInterrupt)
        aiSpeakingEmitted = false
        pushEvent(["type": "interrupted"])
        result(nil)
    }

    // MARK: - EventChannel

    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    private func pushEvent(_ data: [String: Any]) {
        DispatchQueue.main.async { [weak self] in self?.eventSink?(data) }
    }
}

// MARK: - SpeechEngineDelegate

extension VoiceDialogPlugin: SpeechEngineDelegate {

    /// 兼容多种字段名，拿到 ASR 文本：
    /// 1) results: [{text: "..."}, ...]
    /// 2) result.text / text / content
    private func extractAsrText(_ json: [String: Any]) -> String {
        if let arr = json["results"] as? [[String: Any]] {
            let joined = arr.compactMap { $0["text"] as? String }.joined()
            if !joined.isEmpty { return joined }
        }
        if let result = json["result"] as? [String: Any],
           let text = result["text"] as? String, !text.isEmpty {
            return text
        }
        if let text = json["text"] as? String, !text.isEmpty { return text }
        if let content = json["content"] as? String, !content.isEmpty { return content }
        return ""
    }

    /// 兼容 LLM 增量的字段名：content / text / delta
    private func extractChatDelta(_ json: [String: Any]) -> String {
        if let content = json["content"] as? String, !content.isEmpty { return content }
        if let text    = json["text"]    as? String, !text.isEmpty    { return text }
        if let delta   = json["delta"]   as? String, !delta.isEmpty   { return delta }
        return ""
    }

    func onMessage(with type: SEMessageType, andData data: Data) {
        // 部分消息体不是 JSON（空 data 或二进制），此时 json 为空字典
        let json: [String: Any] =
            (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]

        switch type {
        // ── 引擎级（StartEngine 模式）────────────────────────────────────────
        case SEEngineStart:
            NSLog("[VoiceDialog] ENGINE_START")
            engineStarted = true
            aiSpeakingEmitted = false
            pushEvent(["type": "connected"])

        case SEEngineStop:
            NSLog("[VoiceDialog] ENGINE_STOP")
            engineStarted = false
            aiSpeakingEmitted = false
            pushEvent(["type": "disconnected"])

        case SEEngineError:
            let err = (String(data: data, encoding: .utf8)?.nilIfEmpty)
                ?? (json["message"] as? String)
                ?? "Unknown engine error"
            NSLog("[VoiceDialog] ENGINE_ERROR: \(err)")
            pushEvent(["type": "error", "error": err])

        // ── ASR ───────────────────────────────────────────────────────────
        case SEEventASRInfo:
            let text = extractAsrText(json)
            if !text.isEmpty {
                pushEvent(["type": "userSpeaking", "text": text])
            }

        case SEEventASRResponse:
            let text = extractAsrText(json)
            if !text.isEmpty {
                pushEvent(["type": "userFinalText", "text": text])
            }

        case SEEventASREnded:
            aiSpeakingEmitted = false

        // ── Chat（LLM 流式）─────────────────────────────────────────────────
        case SEEventChatResponse:
            let delta = extractChatDelta(json)
            if !delta.isEmpty {
                if !aiSpeakingEmitted {
                    aiSpeakingEmitted = true
                    pushEvent(["type": "aiSpeaking"])
                }
                pushEvent(["type": "aiTextDelta", "text": delta])
            }

        case SEEventChatEnded:
            aiSpeakingEmitted = false

        // ── TTS ───────────────────────────────────────────────────────────
        case SEEventTTSSentenceStart:
            if !aiSpeakingEmitted {
                aiSpeakingEmitted = true
                pushEvent(["type": "aiSpeaking"])
            }

        case SEEventTTSSentenceEnd:
            break

        case SEEventTTSEnded:
            pushEvent(["type": "aiRoundDone"])
            aiSpeakingEmitted = false

        // ── 兼容旧事件（使用 Event 模式时的遗留回调，不影响 Engine 模式）───────
        case SEEventConnectionFailed:
            let err = json["message"] as? String ?? "Connection failed"
            pushEvent(["type": "error", "error": err])

        case SEEventSessionFailed:
            let err = json["message"] as? String ?? "Session failed"
            pushEvent(["type": "error", "error": err])

        default:
            // 未覆盖的事件仅打印用于调试
            NSLog("[VoiceDialog] unhandled msg type=\(type.rawValue) len=\(data.count)")
        }
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

#else

/// 未链接 SpeechEngineToB 时（例如模拟器 Pod 跳过）的占位实现。
class VoiceDialogPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    static let methodChannelName = "com.starpath/voice_dialog"
    static let eventChannelName  = "com.starpath/voice_dialog_events"

    private var eventSink: FlutterEventSink?

    static func register(with registrar: FlutterPluginRegistrar) {
        let instance = VoiceDialogPlugin()
        let methodChannel = FlutterMethodChannel(
            name: methodChannelName,
            binaryMessenger: registrar.messenger()
        )
        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        let eventChannel = FlutterEventChannel(
            name: eventChannelName,
            binaryMessenger: registrar.messenger()
        )
        eventChannel.setStreamHandler(instance)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "prepareEnvironment":
            result(nil)
        case "startDialog":
            result(FlutterError(
                code: "NO_NATIVE_SDK",
                message: "火山语音 SDK 未集成（模拟器开发模式）。真机请删除 ios/.skip_volc_for_sim 后执行 pod install。",
                details: nil
            ))
        case "stopDialog", "interrupt":
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    func onListen(withArguments arguments: Any?,
                  eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

#endif

File: app/android/app/src/main/kotlin/com/example/starpath/VoiceDialogPlugin.kt
package com.example.starpath

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

import com.bytedance.speech.speechengine.SpeechEngine
import com.bytedance.speech.speechengine.SpeechEngineDefines
import com.bytedance.speech.speechengine.SpeechEngineGenerator

/**
 * Flutter ↔ 火山引擎 SpeechEngine Dialog SDK 桥接（Android）
 *
 * 官方文档: https://www.volcengine.com/docs/6561/1597643
 *
 * 正确调用顺序（通过 .class 反编译 SpeechEngineDefines 确认）：
 *   1. PrepareEnvironment
 *   2. getInstance / createEngine / setOption* / setContext / setListener / initEngine
 *   3. sendDirective(DIRECTIVE_DIALOG_START_CONNECTION, "")   ← 建立 WebSocket
 *   4. 回调 MESSAGE_TYPE_DIALOG_CONNECTION_STARTED
 *      → sendDirective(DIRECTIVE_DIALOG_START_SESSION, sessionJson)  ← 启动会话
 *   5. 回调 MESSAGE_TYPE_DIALOG_SESSION_STARTED → 推送 "connected"
 *
 * 错误 -700 (ERR_SEND_DIRECTIVE_IN_WRONG_STATE) 原因：
 *   旧代码误用了 DIRECTIVE_START_ENGINE（通用引擎指令），
 *   Dialog SDK 需要使用 DIRECTIVE_DIALOG_START_CONNECTION。
 */
class VoiceDialogPlugin(
    private val context: Context,
    private val activity: Activity?,
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    companion object {
        private const val METHOD_CHANNEL = "com.starpath/voice_dialog"
        private const val EVENT_CHANNEL  = "com.starpath/voice_dialog_events"

        private const val DIALOG_ADDRESS = "wss://openspeech.bytedance.com"
        private const val DIALOG_URI     = "/api/v3/realtime/dialogue"

        private const val TAG = "VoiceDialogPlugin"

        /** X-Api-App-Key 固定值（官方文档 1594356，非控制台 Secret Key） */
        private const val FIXED_APP_KEY = "PlgvMymc7f3tQnJ6"

        private const val REQUEST_RECORD_AUDIO = 7001
    }

    private var pendingStartArgs:   MethodCall?            = null
    private var pendingStartResult: MethodChannel.Result?  = null

    private val methodChannel = MethodChannel(messenger, METHOD_CHANNEL)
    private val eventChannel  = EventChannel(messenger, EVENT_CHANNEL)
    private var eventSink: EventChannel.EventSink? = null
    private var engine: SpeechEngine? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    private var aiSpeakingEmitted = false

    private val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    private var audioFocusRequest: AudioFocusRequest? = null

    init {
        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(this)
    }

    // ── MethodChannel ─────────────────────────────────────────────────────────

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "prepareEnvironment" -> handlePrepareEnvironment(result)
            "startDialog"        -> handleStartDialog(call, result)
            "stopDialog"         -> handleStopDialog(result)
            "interrupt"          -> handleInterrupt(result)
            else                 -> result.notImplemented()
        }
    }

    private fun handlePrepareEnvironment(result: MethodChannel.Result) {
        try {
            SpeechEngineGenerator.PrepareEnvironment(
                context,
                context.applicationContext as android.app.Application,
            )
            result.success(null)
        } catch (e: Exception) {
            result.error("PREPARE_ENV_ERROR", e.message, null)
        }
    }

    /** MainActivity.onRequestPermissionsResult → 转发到此处 */
    fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode != REQUEST_RECORD_AUDIO) return
        val granted = grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED
        val call   = pendingStartArgs   ?: return
        val result = pendingStartResult ?: return
        pendingStartArgs   = null
        pendingStartResult = null
        if (granted) doStartDialog(call, result)
        else result.error("MIC_DENIED", "麦克风权限被拒绝，无法启动语音对话", null)
    }

    private fun handleStartDialog(call: MethodCall, result: MethodChannel.Result) {
        val appId    = call.argument<String>("appId")    ?: ""
        val appToken = call.argument<String>("appToken") ?: ""
        if (appId.isEmpty() || appToken.isEmpty()) {
            return result.error("MISSING_PARAM", "appId 和 appToken 不能为空", null)
        }

        val hasMic = ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) ==
                PackageManager.PERMISSION_GRANTED
        if (!hasMic) {
            if (activity != null) {
                pendingStartArgs   = call
                pendingStartResult = result
                ActivityCompat.requestPermissions(
                    activity, arrayOf(Manifest.permission.RECORD_AUDIO), REQUEST_RECORD_AUDIO,
                )
            } else {
                result.error("MIC_DENIED", "无法请求麦克风权限（activity 为 null）", null)
            }
            return
        }
        doStartDialog(call, result)
    }

    private fun doStartDialog(call: MethodCall, result: MethodChannel.Result) {
        val appId       = call.argument<String>("appId")       ?: ""
        val appToken    = call.argument<String>("appToken")    ?: ""
        val resourceId  = call.argument<String>("resourceId")  ?: "volc.speech.dialog"
        val dialogModel = call.argument<String>("dialogModel") ?: "1.2.1.1"
        val ttsSpeaker  = call.argument<String>("ttsSpeaker")  ?: "zh_female_vv_jupiter_bigtts"
        val enableAec   = call.argument<Boolean>("enableAec")  ?: false

        android.util.Log.i(TAG, "doStartDialog appId=$appId resourceId=$resourceId model=$dialogModel")

        try {
            requestAudioFocus()

            // 已有引擎先销毁
            engine?.let {
                try { it.sendDirective(SpeechEngineDefines.DIRECTIVE_SYNC_STOP_ENGINE, "") } catch (_: Exception) {}
                try { it.destroyEngine() } catch (_: Exception) {}
                engine = null
            }

            engine = SpeechEngineGenerator.getInstance()
            engine!!.createEngine()

            // ── 必填参数 ────────────────────────────────────────────────────
            engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_ENGINE_NAME_STRING, SpeechEngineDefines.DIALOG_ENGINE)
            engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_APP_ID_STRING,      appId)
            engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_APP_KEY_STRING,     FIXED_APP_KEY)
            engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_APP_TOKEN_STRING,   appToken)
            engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_RESOURCE_ID_STRING, resourceId)
            engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_UID_STRING,         "starpath-user")
            engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_DIALOG_ADDRESS_STRING, DIALOG_ADDRESS)
            engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_DIALOG_URI_STRING,     DIALOG_URI)

            // AEC
            if (enableAec) {
                val aecPath = copyAecModelToFilesDir()
                if (aecPath != null) {
                    engine!!.setOptionBoolean(SpeechEngineDefines.PARAMS_KEY_ENABLE_AEC_BOOL, true)
                    engine!!.setOptionString(SpeechEngineDefines.PARAMS_KEY_AEC_MODEL_PATH_STRING, aecPath)
                    android.util.Log.i(TAG, "AEC 已启用: $aecPath")
                }
            }

            // setContext / setListener 必须在 initEngine 之前（否则早期回调丢失，导致 -700）
            engine!!.setContext(context.applicationContext)
            engine!!.setListener { type, data, len -> handleSpeechMessage(type, data, len) }

            val initRet = engine!!.initEngine()
            if (initRet != SpeechEngineDefines.ERR_NO_ERROR) {
                engine!!.destroyEngine(); engine = null
                return result.error("INIT_ERROR", "initEngine returned $initRet", null)
            }

            // 按官方文档：先 SYNC_STOP，再 START_ENGINE（避免内部异步线程问题）
            engine!!.sendDirective(SpeechEngineDefines.DIRECTIVE_SYNC_STOP_ENGINE, "")

            val startPayload = buildStartEnginePayload(dialogModel, ttsSpeaker)
            android.util.Log.i(TAG, "START_ENGINE payload=$startPayload")
            val startRet = engine!!.sendDirective(SpeechEngineDefines.DIRECTIVE_START_ENGINE, startPayload)
            android.util.Log.i(TAG, "START_ENGINE returned $startRet")
            if (startRet != SpeechEngineDefines.ERR_NO_ERROR) {
                engine!!.destroyEngine(); engine = null
                return result.error("START_ERROR", "sendDirective(START_ENGINE) returned $startRet", null)
            }

            // connected / error 事件在 MESSAGE_TYPE_ENGINE_START 回调里异步推送
            result.success(true)

        } catch (e: Exception) {
            engine?.let { try { it.destroyEngine() } catch (_: Exception) {} }
            engine = null
            result.error("START_ERROR", e.message, null)
        }
    }

    private fun handleStopDialog(result: MethodChannel.Result) {
        try {
            engine?.sendDirective(SpeechEngineDefines.DIRECTIVE_SYNC_STOP_ENGINE, "")
            engine?.destroyEngine()
            engine = null
            abandonAudioFocus()
            result.success(null)
        } catch (e: Exception) {
            result.error("STOP_ERROR", e.message, null)
        }
    }

    private fun handleInterrupt(result: MethodChannel.Result) {
        try {
            engine?.sendDirective(SpeechEngineDefines.DIRECTIVE_EVENT_CLIENT_INTERRUPT, "")
            result.success(null)
        } catch (e: Exception) {
            result.error("INTERRUPT_ERROR", e.message, null)
        }
    }

    // ── 消息回调 ──────────────────────────────────────────────────────────────

    private fun handleSpeechMessage(type: Int, data: ByteArray, len: Int) {
        val strData = String(data, 0, len, Charsets.UTF_8)
        android.util.Log.d(TAG, "▶ onSpeechMessage type=$type len=$len data=${strData.take(200)}")
        when (type) {

            // ── 引擎级（START_ENGINE 模式）────────────────────────────────────
            SpeechEngineDefines.MESSAGE_TYPE_ENGINE_START -> {
                android.util.Log.i(TAG, "ENGINE_START: $strData")
                pushEvent(mapOf("type" to "connected"))
            }
            SpeechEngineDefines.MESSAGE_TYPE_ENGINE_STOP -> {
                android.util.Log.i(TAG, "ENGINE_STOP: $strData")
                pushEvent(mapOf("type" to "disconnected"))
            }
            SpeechEngineDefines.MESSAGE_TYPE_ENGINE_ERROR -> {
                android.util.Log.e(TAG, "ENGINE_ERROR: $strData")
                pushEvent(mapOf("type" to "error", "error" to strData))
            }

            // ── ASR ───────────────────────────────────────────────────────────
            SpeechEngineDefines.MESSAGE_TYPE_DIALOG_ASR_INFO -> {
                val text = parseAsrText(strData).ifEmpty { parseJsonField(strData, "text") ?: "" }
                if (text.isNotEmpty()) pushEvent(mapOf("type" to "userSpeaking", "text" to text))
            }
            SpeechEngineDefines.MESSAGE_TYPE_DIALOG_ASR_RESPONSE -> {
                android.util.Log.i(TAG, "ASR_RESPONSE: $strData")
                val text = parseAsrText(strData)
                if (text.isNotEmpty()) pushEvent(mapOf("type" to "userFinalText", "text" to text))
            }
            SpeechEngineDefines.MESSAGE_TYPE_DIALOG_ASR_ENDED -> {
                android.util.Log.i(TAG, "ASR_ENDED")
                aiSpeakingEmitted = false
            }

            // ── Chat（LLM 流式） ───────────────────────────────────────────────
            SpeechEngineDefines.MESSAGE_TYPE_DIALOG_CHAT_RESPONSE -> {
                val text = parseJsonField(strData, "content") ?: parseJsonField(strData, "text") ?: ""
                android.util.Log.d(TAG, "CHAT_RESPONSE text=$text")
                if (text.isNotEmpty()) {
                    if (!aiSpeakingEmitted) {
                        aiSpeakingEmitted = true
                        pushEvent(mapOf("type" to "aiSpeaking"))
                    }
                    pushEvent(mapOf("type" to "aiTextDelta", "text" to text))
                }
            }
            SpeechEngineDefines.MESSAGE_TYPE_DIALOG_CHAT_ENDED -> {
                android.util.Log.i(TAG, "CHAT_ENDED")
                aiSpeakingEmitted = false
            }

            // ── TTS ───────────────────────────────────────────────────────────
            SpeechEngineDefines.MESSAGE_TYPE_DIALOG_TTS_SENTENCE_START -> {
                if (!aiSpeakingEmitted) {
                    aiSpeakingEmitted = true
                    pushEvent(mapOf("type" to "aiSpeaking"))
                }
            }
            SpeechEngineDefines.MESSAGE_TYPE_DIALOG_TTS_SENTENCE_END -> { /* 单句播完，无需处理 */ }
            SpeechEngineDefines.MESSAGE_TYPE_DIALOG_TTS_ENDED -> {
                android.util.Log.i(TAG, "TTS_ENDED")
                pushEvent(mapOf("type" to "aiRoundDone"))
                aiSpeakingEmitted = false
            }

            // ── 通用兜底 ──────────────────────────────────────────────────────
            SpeechEngineDefines.MESSAGE_TYPE_ENGINE_ERROR -> {
                android.util.Log.e(TAG, "ENGINE_ERROR: $strData")
                pushEvent(mapOf("type" to "error", "error" to strData))
            }
            else -> {
                android.util.Log.d(TAG, "MSG type=$type len=$len data=$strData")
            }
        }
    }

    // ── 工具函数 ──────────────────────────────────────────────────────────────

    /**
     * START_ENGINE payload（官方文档示例）：
     * - dialog.bot_name：人设名，默认豆包
     * - dialog.extra.model：模型版本（O2.0 → 1.2.1.1，SC2.0 → 2.2.0.0）
     * - asr.extra / tts.audio_config 不可为 null
     */
    private fun buildStartEnginePayload(dialogModel: String, speaker: String): String =
        """{"dialog":{"bot_name":"豆包","extra":{"model":"$dialogModel"}},"asr":{"extra":{}},"tts":{"audio_config":{},"speaker":"$speaker"}}"""

    private fun requestAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val attrs = AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_VOICE_COMMUNICATION)
                .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                .build()
            val req = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
                .setAudioAttributes(attrs)
                .setAcceptsDelayedFocusGain(false)
                .setWillPauseWhenDucked(false)
                .setOnAudioFocusChangeListener { }
                .build()
            audioFocusRequest = req
            audioManager.requestAudioFocus(req)
        } else {
            @Suppress("DEPRECATION")
            audioManager.requestAudioFocus(null, AudioManager.STREAM_VOICE_CALL,
                AudioManager.AUDIOFOCUS_GAIN_TRANSIENT)
        }
    }

    private fun abandonAudioFocus() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            audioFocusRequest?.let { audioManager.abandonAudioFocusRequest(it) }
            audioFocusRequest = null
        } else {
            @Suppress("DEPRECATION")
            audioManager.abandonAudioFocus(null)
        }
    }

    private fun copyAecModelToFilesDir(): String? {
        return try {
            val destFile = context.filesDir.resolve("aec/aec.model")
            if (!destFile.exists()) {
                destFile.parentFile?.mkdirs()
                context.assets.open("aec/aec.model").use { it.copyTo(destFile.outputStream()) }
            }
            destFile.absolutePath
        } catch (_: Exception) { null }
    }

    private fun parseAsrText(json: String): String {
        return try {
            val obj = org.json.JSONObject(json)
            val results = obj.optJSONArray("results")
            if (results != null && results.length() > 0) {
                (0 until results.length()).joinToString("") { results.getJSONObject(it).optString("text", "") }
            } else {
                obj.optString("text", "")
            }
        } catch (_: Exception) { "" }
    }

    private fun parseJsonField(json: String, key: String): String? {
        return try {
            val obj = org.json.JSONObject(json)
            if (obj.has(key)) obj.optString(key).takeIf { it.isNotEmpty() } else null
        } catch (_: Exception) { null }
    }

    // ── EventChannel ──────────────────────────────────────────────────────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) { eventSink = events }
    override fun onCancel(arguments: Any?) { eventSink = null }

    private fun pushEvent(data: Map<String, Any?>) {
        mainHandler.post { eventSink?.success(data) }
    }

    fun dispose() {
        try {
            engine?.sendDirective(SpeechEngineDefines.DIRECTIVE_SYNC_STOP_ENGINE, "")
            engine?.destroyEngine()
        } catch (_: Exception) {}
        engine = null
        abandonAudioFocus()
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }
}
